### DISCUSSION TITLE
How to get access to tcfapi?

### DISCUSSION THREAD

**Discussion by rforster-dev** [cypress-io/cypress#26835] [Questions and Help]
We're using Sourcepoint as our CMP.

One thing we are trying to do is make sure that our Adverts are being injected onto a page through Cypress.

So I have gotten round writting a command that accepts the CMP popup, but what we notice is that the `__tcfapi` is not defined, even though we have it on the page.

Has anyone else come across this, and how have they got round it?

**Comment by portyu9** [comment]
Hi @rforster-dev 

The likely issue is **which window you are accessing**.

Cypress test/support code does not run in exactly the same window as the application under test. If Sourcepoint installs the TCF API as:

```js
window.__tcfapi
```

on your application page, accessing bare `__tcfapi` from the Cypress spec context can therefore be `undefined`.

Access it through the AUT window instead:

```js
cy.window()
  .should('have.property', '__tcfapi')
  .then((win) => {
    win.__tcfapi('addEventListener', 2, (tcData, success) => {
      expect(success).to.eq(true)

      // assertions against tcData here
    })
  })
```

Using the callback on `win.__tcfapi(...)` is preferable to trying to reach into the CMP's iframe directly. CMPs commonly communicate with their iframe internally, and that iframe may also be cross-origin.

If this assertion never passes:

```js
cy.window().should('have.property', '__tcfapi')
```

then verify in the application itself that Sourcepoint has installed the API on the **top-level application window** and that CMP initialization has completed. In that situation the problem is no longer Cypress variable scoping; the API simply is not present on the page window Cypress is testing.

I reviewed the discussion and there were no screenshots or follow-up comments providing evidence of a different failure mode.

Please mark this answer as accepted if it helped, thank you.
